src/_gettsim/interface.py

```diff
@@ -2,661 +2,453 @@
     collected_conversions = [
-        "The data types of the following input variables have been converted: \n"
+        "The data types of the following input variables have been converted:"
     ]
     general_warning = (
         "Note that the automatic conversion of data types is unsafe and that"
         " its correctness cannot be guaranteed."
         " The best solution is to convert all columns to the expected data"
         " types yourself."
     )
-    for column_name, series in data.items():
-        # Find out if internal_type is defined
+
+    data_tree_paths = optree.tree_accessors(data_tree)
+    data_tree_with_correct_types = {}
+
+    for accessor in data_tree_paths:
+        qualified_column_name = QUALIFIED_NAME_SEPARATOR.join(accessor.path)
+        data_leaf = accessor(data_tree)
         internal_type = None
-        if column_name in TYPES_INPUT_VARIABLES:
-            internal_type = TYPES_INPUT_VARIABLES[column_name]
-        elif (
-            column_name in functions_overridden
-            and "return" in functions_overridden[column_name].__annotations__
-        ):
-            func = functions_overridden[column_name]
-            if hasattr(func, "__info__") and func.__info__["skip_vectorization"]:
+
+        # Look for column in TYPES_INPUT_VARIABLES
+        try:  # noqa: SIM105ga
+            internal_type = accessor(TYPES_INPUT_VARIABLES)
+        except KeyError:
+            pass
+
+        # Look for column in functions_tree_overridden
+        try:
+            func = accessor(functions_tree_overridden)
+            if hasattr(func, "__annotations__") and func.skip_vectorization:
                 # Assumes that things are annotated with numpy.ndarray([dtype]), might
                 # require a change if using proper numpy.typing. Not changing for now
                 # as we will likely switch to JAX completely.
                 internal_type = get_args(func.__annotations__["return"])[0]
-            elif func in create_groupings().values():
-                # Functions that create a grouping ID
-                internal_type = get_args(func.__annotations__["return"])[0]
             else:
                 internal_type = func.__annotations__["return"]
+        except KeyError:
+            pass
 
         # Make conversion if necessary
-        if internal_type and not check_series_has_expected_type(series, internal_type):
+        if internal_type and not check_series_has_expected_type(
+            series=data_leaf, internal_type=internal_type
+        ):
             try:
-                data[column_name] = convert_series_to_internal_type(
-                    series, internal_type
+                converted_leaf = convert_series_to_internal_type(
+                    series=data_leaf, internal_type=internal_type
+                )
+                data_tree_with_correct_types = upsert_path_and_value(
+                    tree=data_tree_with_correct_types,
```

Comment:
```suggestion
                    base=data_tree_with_correct_types,
```
match interface of `upsert_tree`.